[Merged by Bors] - feat: add `compile_inductive%` and `compile_def%` commands

gebner opened the pull request:
Add a `#compile inductive` command to compile the recursors of an inductive type, which works by creating a recursive definition and using `@[csimp]`.

---

A revival of #1806 with a branch on the mathlib4 repo.  cc @cppio @rwbarton from the old PR

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

digama0 commented:
Could you summarize how this PR differs from the previous one?

digama0 commented on Mathlib/Util/CompileInductive.lean:
```diff
@@ -0,0 +1,196 @@
+for which Lean does not generate compiled code by default
+(since it is not used 99% of the time).
+-/
+elab tk:"#compile " "def " i:ident : command => Command.liftTermElabM do
```
This shouldn't have a `#` on the name, because it is not a temporary use command. Rather, I think it should have a `%` suffix similar to `gen_injective_theorems%` to reflect that it is somewhat internal-use and also a workaround for a current lean limitation.

gebner replied:
It's `compile_def%` and `compile_inductive%` now.

gebner commented:
> Could you summarize how this PR differs from the previous one?

 - Merged master
 - Addressed all review comments from #1806
 - Add go-to-definition support for the identifier
 - Added a bunch of `#compile` commands, removing noncomputable porting notes along the way
 - Removed `import Mathlib` from the test

The reason I made a new one is because I didn't want to push to @cppio's master branch.

digama0 commented:
Last time I checked, Leo had a concern regarding compiled recursors that the type signature does not reflect the correct laziness / order of operations. Could you add documentation and/or a test which indicates what evaluation order these inductives have? My guess is that e.g. `List.rec` will evaluate the nil case unconditionally and all the other cases from the bottom up, although you can work around this by using the recursor at type `Unit -> A` instead of `A`.

gebner commented:
I've added a warning to the file.  I'm not sure if `Unit → A` would help, since inlining happens after replacing the recursor.

digama0 commented:
Here's a demonstration:
```

def matchFoo : ℕ → ℕ
  | 0 => dbg_trace "Z"; 0
  | a + 1 => dbg_trace "A:{a}"; (dbg_trace "B:{a}"; matchFoo a) + (dbg_trace "C:{a}"; 1)

def recFoo : ℕ → ℕ :=
fun n => @Nat.rec (fun _ => ℕ) (dbg_trace "Z"; 0)
  (fun a as => dbg_trace "A:{a}"; (dbg_trace "B:{a}"; as) + (dbg_trace "C:{a}"; 1)) n

def recUnitFoo : ℕ → ℕ :=
fun n => @Nat.rec (fun _ => Unit → ℕ) (fun _ => dbg_trace "Z"; 0)
  (fun a as _ => dbg_trace "A:{a}"; (dbg_trace "B:{a}"; as ()) + (dbg_trace "C:{a}"; 1)) n ()

#eval matchFoo 5   -- A:4 B:4 A:3 B:3 A:2 B:2 A:1 B:1 A:0 B:0 Z C:0 C:1 C:2 C:3 C:4
#eval recFoo 5     -- Z A:0 B:0 C:0 A:1 B:1 C:1 A:2 B:2 C:2 A:3 B:3 C:3 A:4 B:4 C:4
#eval recUnitFoo 5 -- A:4 B:4 A:3 B:3 A:2 B:2 A:1 B:1 A:0 B:0 Z C:0 C:1 C:2 C:3 C:4
```
* `matchFoo` is what you get with the equation compiler, this is the desired evaluation order, evaluating `A:4 B:4`, then the n=3 recursive call, then `C:4`.
* `recFoo` is what is implemented in this PR with a basic use of `Nat.rec`: we first evaluate the n=3 recursive call, then `A:4 B:4 C:4` at the end (resulting in an overall bottom up evaluation order).
* `recUnitFoo` is the workaround: use `Nat.rec` as in this PR, but with a motive of `Unit -> Nat` instead, and introduce a `fun _ =>` in each inductive case and a `()` on each occurrence of the IH in the closure. This restores the desired evaluation order, but it is unlikely we can do this transformation automatically unless we have access to the expression itself and some kind of macro expansion. We could still use it manually, however.

digama0 commented on Mathlib/Util/CompileInductive.lean:
```diff
@@ -0,0 +1,204 @@
+    name
+    levelParams := rv.levelParams
+    type := ← mkEq rv.type old new
+    value := ← forallTelescope rv.type λ xs body => do
+      let pf := .const rv.name (.zero :: levels.tail!)
+      let pf := mkAppN pf xs[:rv.numParams]
+      let old := mkAppN old xs
+      let new := mkAppN new xs
+      let motive ← mkLambdaFVars xs[rv.getFirstIndexIdx:] <| ← mkEq body old new
+      let pf := .app pf motive
+      let pf := mkAppN pf <| ← rv.rules.toArray.zip xs[rv.getFirstMinorIdx:]
+        |>.mapM λ (rule, minor) => do
+        forallTelescope ((← inferType minor).replaceFVar xs[rv.numParams]! motive) λ ys _ => do
+          let minor := mkAppN minor ys[:rule.nfields]
+          let pf' ← mkEqRefl (← inferType minor) minor
+          let pf' ← ys[rule.nfields:].foldlM (λ pf' y => do mkCongr pf' (← mkFunExts y)) pf'
+          mkLambdaFVars ys pf'
+      let pf := mkAppN pf xs[rv.getFirstIndexIdx:]
+      mkFunExts' xs pf (body, old, new)
```
```suggestion
    value := .const name (rv.levelParams.map .param)
```
Not sure how seriously to take this as a suggestion, but it works... Since this declaration is declared as `safety := .partial` it's already working in an inconsistent context and we don't have to prove anything.

gebner replied:
Oh! I thought that the kernel checks the inhabitedness of the type.  Because the usual story is "partial doesn't compromise soundness".

In that case we might even do away with the csimp charade entirely and just declare it as implemented_by.

digama0 replied:
Yeah, `DefinitionSafety.partial` is really confusing, because it's not the same thing as the `partial` modifier users can put on definitions. `DefinitionSafety.partial` is basically a synonym of "unsafe" as far as the kernel is concerned: you are allowed to do general recursion, but you cannot directly reference it in safe declarations. When you write `partial def foo : A := foo`, lean does `addAndCompile { name := foo', safety := .partial, value := foo' }` to declare the computational version and `addDecl { name := foo, safety := .safe, value := default }` for the model definition, and then connects the two with `implemented_by`.

gebner replied:
Apparently this is just inconsistent terminology.  Kernel-level partial defs are different than user-level partial defs.  The kernel ones are essentially the same as unsafe, while the user-level ones are perfectly safe and usable in proofs.

```lean
import Lean
open Lean

#eval addDecl <| .mutualDefnDecl [{
  name := `False.intro
  levelParams := []
  type := .const ``False []
  value := .const `False.intro []
  hints := .opaque
  safety := .partial
}]

#check False.intro -- False
#print axioms False.intro -- 'False.intro' does not depend on any axioms

theorem please_dont_typecheck : False := False.intro
-- (kernel) invalid declaration, safe declaration must not contain partial declaration 'False.intro'
```

gebner replied:
FWIW, `implemented_by` doesn't work across modules.  So let's go with your unsafe proof version.

digama0 replied:
> ```lean
> #print axioms False.intro -- 'False.intro' does not depend on any axioms
> ```

We should really mention when a constant is `partial` or `unsafe` in `#print axioms`...

digama0 replied:
Another variation would be to use `lcProof` and make the equality `.unsafe`. But I guess this version results in a smaller proof term.

gebner commented:
!bench

digama0 commented:
I've contributed significantly to this PR so I'm not sure I am a good reviewer at this point, but the perf results look good (ignore the wall of red, which is comparing against the wrong commit, the bit at the end is the correct comparison and it is basically perf-neutral), the implementation looks good to me, and this issue keeps coming up in the port.

bors r+
